src/prommis/nanofiltration/diafiltration.py

```diff
@@ -0,0 +1,774 @@
+        )
+    )
+    m.Li_purity = Expression(
+        expr=m.fs.stage3.permeate_outlet.flow_vol[0]
+        * m.fs.stage3.permeate_outlet.conc_mass_solute[0, "Li"]
+        / (
+            (
+                m.fs.stage3.permeate_outlet.flow_vol[0]
+                * m.fs.stage3.permeate_outlet.conc_mass_solute[0, "Li"]
+            )
+            + (
+                m.fs.stage3.permeate_outlet.flow_vol[0]
+                * m.fs.stage3.permeate_outlet.conc_mass_solute[0, "Co"]
+            )
+        )
+    )
+    m.Co_recovery = Expression(
+        expr=m.fs.stage1.retentate_outlet.flow_vol[0]
+        * m.fs.stage1.retentate_outlet.conc_mass_solute[0, "Co"]
+        / (
+            m.fs.mix2.inlet_1.flow_vol[0] * m.fs.mix2.inlet_1.conc_mass_solute[0, "Co"]
+            + m.fs.stage3.retentate_side_stream_state[0, 10].flow_vol
+            * m.fs.stage3.retentate_side_stream_state[0, 10].conc_mass_solute["Co"]
+        )
+    )
+    m.Co_purity = Expression(
+        expr=m.fs.stage1.retentate_outlet.flow_vol[0]
+        * m.fs.stage1.retentate_outlet.conc_mass_solute[0, "Co"]
+        / (
+            (
+                m.fs.stage1.retentate_outlet.flow_vol[0]
+                * m.fs.stage1.retentate_outlet.conc_mass_solute[0, "Co"]
+            )
+            + (
+                m.fs.stage1.retentate_outlet.flow_vol[0]
+                * m.fs.stage1.retentate_outlet.conc_mass_solute[0, "Li"]
+            )
+        )
+    )
+
+    m.membrane_length = Expression(
+        expr=m.fs.stage1.length + m.fs.stage2.length + m.fs.stage3.length
+    )
+
+
+def solve_model(m):
+    """
+    Method to solve the diafiltration flowsheet
+
+    Args:
+        m: Pyomo model
+    """
+    solver = get_solver()
+    solver.options = {"max_iter": 3000}
+    results = solver.solve(m, tee=True)
+    if results.solver.termination_condition != "optimal":
+        raise ValueError("The solver did not return optimal termination")
+
+    # TODO: add Boolean variable to calculate pump OPEX
+    # Verify the feed pump operating pressure workaround is valid
+    # assume this additional cost is less than half a cent
+    if value(m.fs.feed_pump.costing.variable_operating_cost) >= 0.005:
+        raise ValueError(
+            "The variable   operating cost of the feed pump as calculated in the feed"
+            "pump costing block is not negligible. This operating cost is already"
+            "accounted for via the membrane pressure drop specific energy consumption."
+        )
+
+    return results
+
+
+def add_costing(m):
+    """
+    Method to add costing block to the flowsheet
+
+    Args:
+        m: Pyomo model
+    """
+    # creating dummy variables to store the UnitModelCostingBlocks
+    m.fs.cascade = UnitModelBlock()  # to cost the pressure drop
+    m.fs.feed_pump = UnitModelBlock()  # to cost feed pump
+    m.fs.diafiltrate_pump = UnitModelBlock()  # to cost diafiltrate pump
+
+    m.fs.costing = DiafiltrationCosting()
```

Comment:
Same here, remove this line so that users can define their own flowsheet-level costing block.